src/imitation/density_baselines.py

```diff
@@ -0,0 +1,267 @@
+               is_stationary=False):
+    r"""Family of simple imitation learning baseline algorithms that apply RL to
+    maximise a rough density estimate of the demonstration trajectories.
+    Specifically, it constructs a non-parametric estimate of `p(s)`, `p(s,s')`,
+    `p_t(s,a)`, etc. (depending on options), then rewards the imitation learner
+    with `r_t(s,a,s')=\log p_t(s,a,s')` (or `\log p(s,s')`, or whatever the
+    user wants the model to condition on).
+
+    Args:
+      env (gym.Env or str): environment to train on.
+      imitation_trainer (BaseRLModel): RL algorithm & initial policy that will
+        be used to train the imitation learner.
+      expert_trainer (BaseRLModel): snapshot of RL algorithm that has been
+        trained to solve `env`.
+      n_expert_trajectories (int): number of expert trajectories to sample and
+        use as demonstrations.
+      kernel, kernel_bandwidth, density_type, is_stationary,
+        n_expert_trajectories: these are passed directly to `DensityReward`;
+        refer to documentation for that class."""
+    self.env = util.maybe_load_env(env, vectorize=True)
+    self.imitation_trainer = imitation_trainer
+    self.reward_fn = DensityReward(trajectories=rollouts,
+                                   density_type=density_type,
+                                   obs_space=self.env.observation_space,
+                                   act_space=self.env.action_space,
+                                   is_stationary=is_stationary,
+                                   kernel=kernel,
+                                   kernel_bandwidth=kernel_bandwidth,
+                                   standardise_inputs=standardise_inputs)
+    self.wrapped_env = reward_wrapper.RewardVecEnvWrapper(self.env,
+                                                          self.reward_fn,
+                                                          include_steps=True)
+    self.graph = tf.Graph()
+    self.sess = tf.Session(graph=self.graph)
+    with self.graph.as_default():
+      self.sess.run(tf.global_variables_initializer())
+
+  def train_policy(self, n_timesteps=int(1e6), **kwargs):
+    """Train the imitation policy for a given number of timesteps. Does not
+    return anything.
+
+    Args:
+      n_timesteps (int): number of timesteps to train the policy for.
+      kwargs (dict): extra arguments that will be passed to the `learn()`
+        method of the imitation RL model. Refer to Stable Baselines docs for
+        details."""
+    self.imitation_trainer.set_env(self.wrapped_env)
+    # FIXME: learn() is not meant to be called frequently; there are
+    # significant per-call overheads (see Adam's comment in trainer.py)
+    # FIXME: the ep_reward_mean reported by SB is wrong; it comes from a
+    # Monitor() that is being (incorrectly) used to wrap the underlying
+    # environment.
+    self.imitation_trainer.learn(n_timesteps,
+                                 reset_num_timesteps=False,
```

Comment:
Why is this set to False rather than letting `kwargs` decide? I'm not sure why this is important. (maybe leave a comment)

Reply:
So that we can see the total number of timesteps of training. Without an appropriate value, the timesteps shown in the log don't reset to 0 each time `.learn` is called. Will comment.